src/interface/isosurface.jl

```diff
@@ -0,0 +1,157 @@
+"""
+    isosurface(x, y, z, V; kw...)
+
+Extract and plot isosurface from volumetric data, or implicit function.
+
+# Usage
+
+    isosurface(x, y, z, V; $(keywords((isovalue = 0, centroid = true); add = (Z_DESCRIPTION..., PROJ_DESCRIPTION..., :canvas), remove = (:blend, :grid))))
+
+# Arguments
+
+$(arguments(
+    (
+        V = "`Array` (volume) of interest for which a surface is extracted, or `Function` evaluated as `f(x, y, z)`",
+        isovalue = "chosen surface isovalue",
+        legacy = "use the legacy Marching Cubes algorithm instead of the topology enhanced algorithm",
+        cull = "cull (hide) back faces",
+        centroid = "display triangulation centroid instead of triangle vertices",
+    ); add = (Z_DESCRIPTION..., PROJ_DESCRIPTION..., :x, :y, :z, :canvas), remove = (:blend, :grid)
+))
+
+# Author(s)
+
+- T Bltg (github.com/t-bltg)
+
+# Examples
+
+```julia-repl
+julia> torus(x, y, z, r = .2, R = .5) = (√(x^2 + y^2) - R)^2 + z^2 - r^2
+julia> isosurface(-1:.1:1, -1:.1:1, -1:.1:1, torus; xlim = (-.5, .5), ylim = (-.5, .5), elevation = 50)
+        ┌────────────────────────────────────────┐ 
+    0.5 │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢀⢀⢀⠠⢄⢄⠄⠄⡠⡠⠤⡀⡀⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⠀⠀⠀⠀⠀⡀⠔⠌⠜⠀⠡⠠⠁⠡⠠⠁⠈⠄⠌⠈⠄⠌⠀⠪⠠⠤⡀⡀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⠀⠀⡠⠔⠅⠌⠈⠄⠌⠀⡁⡀⠨⡀⠄⠠⠡⠠⡀⠥⠠⠈⠀⠡⠠⠁⠡⠱⠢⡀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⢀⣜⠃⡁⠄⢊⠈⠄⡰⠐⡀⢂⡞⢈⡰⡨⡂⢆⡁⣱⠔⢁⠘⢀⠠⠁⡑⠠⢈⠘⣲⡀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⡚⠥⠁⡀⠂⢂⠈⠄⡖⡫⠗⠋⠉⠀⠀⠀⠀⠀⠈⠉⠉⠱⢝⠱⠠⠁⡐⠐⢀⠈⢌⢧⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⡟⠤⠁⡀⠂⢂⢈⠄⢮⡇⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⢸⡿⠠⡁⡐⠐⢀⠈⢤⢚⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⢽⠗⠅⡀⠂⢂⢀⠂⢂⠓⡢⡄⣀⠀⠀⠀⠀⠀⢀⡀⢀⢰⠂⡑⠐⡀⡐⠐⠀⠨⢔⡟⠀⠀⠀⠀│ 
+        │⠀⠀⠀⠀⠈⠭⠄⠌⠰⠀⡀⠂⢂⠀⡐⠐⠌⠔⠑⠅⠡⠊⠢⠡⠂⢂⠀⡐⠐⣀⠐⠠⡁⡃⡑⠁⠀⠀⠀⠀│ 
+        │⠀⠀⠀⢀⠀⠀⠈⠨⠅⠄⠌⠀⡆⢀⠐⠐⢀⠐⠐⠀⠄⠂⠂⡀⠂⡂⡀⠂⢂⠀⡂⢐⠔⠈⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⢸⠀⠀⠀⠀⠀⠈⠢⢥⢀⡂⡃⠇⠃⡁⡡⠨⡈⡨⠈⠬⠨⠢⢈⢠⠑⠒⠈⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⠀⢸⠀⠀⠀⠀⠀⠀⠀⠀⠀⠈⠈⠋⠓⠂⠊⠒⠂⠚⠘⠚⠙⠉⠁⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        │⠀⠀⣀⠼⢄⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+   -0.5 │⠔⠉⠀⠀⠀⠈⠑⠤⢀⡀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀│ 
+        └────────────────────────────────────────┘ 
+        ⠀-0.5⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀0.5⠀ 
+```
+
+# See also
+
+`Plot`, `scatterplot`
+"""
+function isosurface(
+    x::AbstractVector,
+    y::AbstractVector,
+    z::AbstractVector,
+    V::Union{Function,AbstractArray};
+    canvas::Type = BrailleCanvas,
+    color::UserColorType = KEYWORDS.color,
+    name::AbstractString = KEYWORDS.name,
+    projection::Union{MVP,Symbol} = KEYWORDS.projection,
+    isovalue::Number = 0,
+    centroid::Bool = true,
+    legacy::Bool = false,
+    cull::Bool = false,
+    kw...,
+)
+    if V isa Function
+        xx = repeat(x', length(y), 1)
+        yy = repeat(y, 1, length(x))
+        X = repeat(xx, 1, 1, length(z))
+        Y = repeat(yy, 1, 1, length(z))
+        Z = zero(X)
+        for (i, zi) in enumerate(z)
+            Z[:, :, i] .= zi
+        end
+        V = map(V, X, Y, Z) |> Array
```

Comment:
Broadcasting is simple and efficient:

```suggestion
        V = V.(y, x', reshape(z, 1, 1, length(z)))
```

```julia
function mapf(f, x, y, z)
    xx = repeat(x', length(y), 1)
    yy = repeat(y, 1, length(x))
    X = repeat(xx, 1, 1, length(z))
    Y = repeat(yy, 1, 1, length(z))
    Z = zero(X)
    for (i, zi) in enumerate(z)
        Z[:, :, i] .= zi
    end
    map(f, X, Y, Z)
end

broadcast_f(f, x, y, z) = f.(y, x', reshape(z, 1, 1, length(z)))
torus(x, y, z, r = .2, R = .5) = (√(x^2 + y^2) - R)^2 + z^2 - r^2
x, y, z = -1:.1:1, -1:.1:1, -1:.1:1

mapf(f, x, y, z) == broadcast_f(f, x, y, y) # true
@btime mapf(f, $x, $y, $z); # 40.129 μs (16 allocations: 297.25 KiB)
@btime broadcast_f(f, $x, $y, $z); # 21.685 μs (4 allocations: 72.78 KiB)
```

Reply:
That's very nice (old habit from `MATLAB` and/or `Numpy` are hard to get rid of). Wait ... `x` and `y` shouldn't be permuted ! Let me have a look.

Reply:
X and Y are permuted in 3D since we use 1st dim = x, 2nd dim = y, 3rd dim = z.
In 2D however, we use the 1st dim = y and the 2nd dim = x (plot conventions).

Reply:
> Maybe we should use 1st dim = z, 2nd dim = y, 3rd dim = x for 3D plots ?

I have no preference here... You can choose whatever makes sense for the broader Plots ecosystem.

Reply:
The 3D convention here makes sense. I'm afraid that if we want to be consistent, I'll have to open a PR to flip axes for 2D plots. Unless it is justified by colon major access, I'll dig into that later.